public/content/developers/tutorials/secret-state/index.md

```diff
@@ -0,0 +1,682 @@
+
+The simplest solution is to use a [server component](/developers/tutorials/server-components/) to hold the secret state. However, the reason we use blockchain is to prevent cheating by the game developer. We need to ensure the server component's honesty. The server can provide a hash of the state, and use [zero-knowledge proofs](/zero-knowledge-proofs/#why-zero-knowledge-proofs-are-important) to prove that the state used to calculate the result of a move is the correct one.
+
+After reading this article you will know how to create this kind of secret state holding server, a client for showing the state, and an onchain component for communication between the two. The main tools we use will be:
+
+- [Zokrates](https://zokrates.github.io/) for zero-knowledge proofs and their verification.
+- [Typescript](https://www.typescriptlang.org/) for both the server and the client.
+- [Node](https://nodejs.org/en) for the server.
+- [Viem](https://viem.sh/) for communication with the Blockchain.
+- [MUD](https://mud.dev/) for onchain data management. Proper disclosure: While I'm writing this on my own time for Ethereum Foundation, my day job is working for [Lattice](https://lattice.xyz/) which makes this framework.
+- [React](https://react.dev/) for the user interface of the client.
+- [Vite](https://vitejs.dev/) for serving the client code.
+
+## Minesweeper example {#minesweeper}
+
+[Minesweeper](<https://en.wikipedia.org/wiki/Minesweeper_(video_game)>) is a game that includes a secret map with a minefield. The player chooses to dig in a specific location. If that location has a mine, it's game over. Otherwise, the player gets the number of mines in the eight squares surrounding that location.
+
+This application is written using [MUD](https://mud.dev/), a framework that lets us store data onchain using a [key-value database](https://aws.amazon.com/nosql/key-value/) and synchronize that data automatically with offchain components. In addition to synchronization, MUD makes it easy to provide access control, and for other users to [extend](https://mud.dev/guides/extending-a-world) our application permissionlessly.
+
+### Running the minesweeper example {#running-minesweeper-example}
+
+To run the minesweeper example:
+
+1. Make sure you [have the prerequisites installed](https://mud.dev/quickstart#prerequisites): [Node](https://mud.dev/quickstart#prerequisites), [Foundry](https://book.getfoundry.sh/getting-started/installation), [`git`](https://git-scm.com/downloads), and [`pnpm`](https://git-scm.com/downloads).
+
+2. Clone the repository.
+
+   ```sh copy
+   git clone https://github.com/qbzzt/20240901-secret-state.git
+   ```
+
+3. Install the packages.
+
+   ```sh copy
+   cd 20240901-secret-state/
+   pnpm install
```

Comment:
Personally encountered some node and pnpm version issues here. Looks like user will need node 18 and pnpm 8 or 9? I had issues above this.

Reply:
I tried with node 20 and pnpm 10, and it worked - after I removed the requirement from `package.json` ;-)